contracts/two_prisoners_dilemma/lib.rs

```diff
@@ -0,0 +1,527 @@
+        result: String,
+    }
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct TwoPrisonersDilemma {
+        /// A reward a cooperating player receives when other one defect
+        bad_luck_payoff: u128,
+
+        /// A reward both players receive when they both defect
+        defect_payoff: u128,
+
+        /// A reward both players receive when they both cooperate
+        cooperation_payoff: u128,
+
+        /// A reward a defecting player gets when other one cooperate
+        temptation_payoff: u128,
+
+        /// Players choices, lookup of account to either choice of cooperate or defect
+        choices: Mapping<AccountId, Choice>,
+
+        /// How many legitimate votes has been made so far
+        /// legitimate: not duplicate, and from registered players only
+        vote_count: u32,
+
+        /// Registered player accounts
+        players: [AccountId; NUMBER_OF_PLAYERS],
+
+        /// Claimed players rewards
+        claimed_rewards: Mapping<AccountId, ()>,
+    }
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// sender is not registered player, either when trying to cooperate or defect
+        CallerIsNotPlayer,
+
+        /// player already chosen, ie cooperated or defected
+        PlayerAlreadyChosen,
+
+        /// not all players made their choices
+        NotAllPlayersMadeTheirChoices,
+
+        /// contract's funds are insufficient to pay out reward to a player
+        InsufficientContractFunds,
+
+        /// requested transfer failed, this can be the case if the contract does not
+        /// have sufficient free funds or if the transfer would have brought the
+        /// contract's balance below minimum balance
+        TransferFailed,
+
+        /// a reward has been already claimed
+        RewardAlreadyClaimed,
+    }
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl TwoPrisonersDilemma {
+        /// Creates new instance of Prisoner's Dilemma SC with below parameters:
+        /// # Arguments
+        ///
+        /// * `players` - list of exactly 2 accounts what will play the game
+        ///
+        /// Game has following rewards configured (in tokens):
+        /// * `bad_luck_payoff`: 0
+        /// * `defect_payoff`: 1
+        /// * `cooperation_payoff`: 2,
+        /// * `temptation_payoff`: 3.
+        #[ink(constructor)]
+        pub fn new(players: [AccountId; 2]) -> Self {
+            initialize_contract(|contract: &mut Self| {
+                contract.players = players.clone();
+                for player in players {
+                    contract.choices.insert(&player, &Choice::Waiting);
+                }
+
+                contract.bad_luck_payoff = 0;
+                contract.defect_payoff = TOKEN;
+                contract.cooperation_payoff = 2 * TOKEN;
+                contract.temptation_payoff = 3 * TOKEN;
```

Comment:
If all these are constants, why do you put them as storage items? It would be cheaper just to do:
```
const BAD_LUCK_PAYOFF = 0;
const DEFECT_PAYOFF = TOKEN;
etc
```

Reply:
My intention was, in the future this SC can have constructor which initilize these values to make the game more general, but I guess for the first version it is fine to hardcode it.